cupy/cuda/memory.pyx

```diff
@@ -462,14 +458,19 @@ cdef class SingleDeviceMemoryPool:
         # find best-fit, or a smallest larger allocation
         length = len(self._free)
         for i in range(index, length):
-            free_list = self._free[i]
-            if free_list:
-                chunk = free_list.pop()
-                chunk, _remaining = self._split(chunk, size)
-                break
-
-        # cudaMalloc if not found
-        if chunk is None:
+            try:
+                rlock.lock_fastrlock(self._free_lock, -1, True)
```

Comment:
This lock does exclusion control with `free_all_blocks`, too.
I am not sure how we can do double-checked locking here.
Can you elaborate more?